### PR TITLE
Reduce use of LocalDefId <-> HirId maps

### PR DESCRIPTION
**Pull request by cjgillot**
This is an attempt to reduce the perf effect of https://github.com/rust-lang/rust/pull/89278.
r? @ghost

### CODE REVIEW

**Comment by cjgillot** [comment]
r? rust-lang/compiler

**Review by davidtwco — APPROVED** [review]
This looks reasonable to me, but probably warrants another opinion.

**Comment by davidtwco** [comment]
r? rust-lang/compiler

**Comment by nagisa** [comment]
Will not be able to take a look at this before 27th. Please reassign if this needs a quicker review.

File: compiler/rustc_middle/src/hir/map/mod.rs [hunk 1]
```diff
@@ -340,11 +340,23 @@ impl<'hir> Map<'hir> {
...
+    }
+
     /// Retrieves the `Node` corresponding to `id`, panicking if it cannot be found.
     pub fn get(&self, id: HirId) -> Node<'hir> {
         self.find(id).unwrap_or_else(|| bug!("couldn't find hir id {} in the HIR map", id))
     }
 
+    /// Retrieves the `Node` corresponding to `id`, panicking if it cannot be found.
+    #[inline]
+    pub fn get_def(&self, id: LocalDefId) -> Node<'hir> {
```
Comment (nagisa) [inline compiler/rustc_middle/src/hir/map/mod.rs:356]: This function name suggests to me “get a def(id)” while what it really means is “get keyed by defid”. Would it be too onerous to make the name more accurate?

File: compiler/rustc_resolve/src/late/lifetimes.rs [hunk 2]
```diff
@@ -2030,32 +2027,31 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
...
-                    debug!("hir id first={:?}", hir_id);
-                    if let Some((id, span, name)) = match self.tcx.hir().get(hir_id) {
-                        Node::Lifetime(hir_lifetime) => Some((
-                            hir_lifetime.hir_id,
-                            hir_lifetime.span,
-                            hir_lifetime.name.ident(),
-                        )),
-                        Node::GenericParam(param) => {
-                            Some((param.hir_id, param.span, param.name.ident()))
+                    debug!("hir id first={:?}", def_id);
```
Comment (nagisa) [inline compiler/rustc_resolve/src/late/lifetimes.rs:2030]: this is no longer printing the hir_id, is it? That said, `debug!(?def_id)` should be more future proof ^^

**Review by nagisa — APPROVED** [review]
Yeah, the idea seems sound. Couple nits inline, up to you if you think its worth to address them or not.

**Comment by nagisa** [comment]
I'm happy to r=me this. Please rebase and squash the history a little before merging though.

**Pull request merged**